src/eigsolve/lanczos.jl

```diff
@@ -150,3 +150,141 @@ function eigsolve(A, x₀, howmany::Int, which::Selector, alg::Lanczos;
+    if howmany > krylovdim
+        error("krylov dimension $(krylovdim) too small to compute $howmany eigenvalues")
+    end
+    tol = alg.tol
+    verbosity = alg.verbosity
+
+    # The initial block size is determined by the number of the starting vectors provided by the user
+    S = typeof(inner(x₀[1], x₀[1])) # Scalar type of the inner product between vectors in x₀
+    block0 = BlockVec{S}(x₀)
+    bs = length(block0)
+    bs < 1 && error("The length of the starting vector x₀ must be greater than 0")
+
+    iter = BlockLanczosIterator(A, block0, krylovdim + bs, alg.qr_tol, alg.orth)
+    fact = initialize(iter; verbosity=verbosity)  # Returns a BlockLanczosFactorization
+    numops = 1    # Number of matrix-vector multiplications (for logging)
+    numiter = 1
+
+    converged = 0
+    local normresiduals, D, U
+
+    while true
+        K = length(fact)
+        β = normres(fact)
+
+        if β < tol && K < howmany && verbosity >= WARN_LEVEL
+            msg = "Invariant subspace of dimension $(K) (up to requested tolerance `tol = $tol`), "
+            msg *= "which is smaller than the number of requested eigenvalues (i.e. `howmany == $howmany`)."
+            @warn msg
+        end
+        # BlockLanczos can access the case of K = 1 and doesn't need extra processing
+        if K >= krylovdim || β <= tol || (alg.eager && K >= howmany)
+            # Compute eigenvalues
+            # Note: Fast eigen solver for block tridiagonal matrices is not implemented yet.
+            BTD = view(fact.T, 1:K, 1:K)
+            D, U = eigen(Hermitian(BTD))
+            by, rev = eigsort(which)
+            p = sortperm(D; by=by, rev=rev)
+            D, U = permuteeig!(D, U, p)
+
+            # Detect convergence by computing the residuals
+            bs_r = fact.r_size   # The block size of the residual (decreases as the iteration goes)
+            r = residual(fact)
+            UU = U[(end - bs_r + 1):end, :]  # The last bs_r rows of U, used to compute the residuals
+            normresiduals = let R = block_inner(r, r)
+                map(u -> sqrt(real(dot(u, R, u))), cols(UU))
+            end
+            converged = count(<=(tol), normresiduals)
+            if converged >= howmany || β <= tol  # Successfully find enough eigenvalues.
+                break
+            elseif verbosity >= EACHITERATION_LEVEL
+                @info "BlockLanczos eigsolve in iteration $numiter: $converged values converged, normres = $(normres2string(normresiduals))"
+            end
+        end
+
+        if K < krylovdim
+            expand!(iter, fact; verbosity=verbosity)
+            numops += 1
+        else # Shrink and restart. Following the shrinking method of Lanczos in the above `eigsolve`.
+            numiter >= maxiter && break
+            bsn = max(div(3 * krylovdim + 2 * converged, 5) ÷ bs, 1) # Divide basis into blocks with the same size
+            keep = bs * bsn
+            H = zeros(S, (bsn + 1) * bs, bsn * bs)
```

Comment:
How about the shorter alternative
```suggestion
            keep = ceil(Int, (3 * krylovdim + 2 * converged) / (5 * bs)) * bs
            H = zeros(S, keep + bs, keep)
```

Reply:
Removing `bsn` is nice! But we should replace `ceil` with `floor`. Otherwise, when `krylovdim=20` and `converged=1`, we get `keep=13`, which is different from `k=12` in shrinking of `Lanczos`.